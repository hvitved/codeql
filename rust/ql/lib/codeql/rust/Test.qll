module Test2 {
  predicate pred(int i) { i = 10 }
}
