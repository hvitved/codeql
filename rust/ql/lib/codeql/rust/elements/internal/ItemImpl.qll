// generated by codegen, remove this comment if you wish to edit this file
/**
 * This module provides a hand-modifiable wrapper around the generated class `Item`.
 *
 * INTERNAL: Do not use.
 */

private import codeql.rust.elements.internal.generated.Item

/**
 * INTERNAL: This module contains the customizable definition of `Item` and should not
 * be referenced directly.
 */
module Impl {
  /**
   * A Item. For example:
   * ```rust
   * todo!()
   * ```
   */
  class Item extends Generated::Item { }

  private import rust
  private import codeql.rust.elements.internal.generated.ParentChild

  private class NamedItem extends Item {
    private string name;

    NamedItem() {
      name = this.(Const).getName().getText()
      or
      name = this.(Enum).getName().getText()
      or
      name = this.(Function).getName().getText()
      or
      name = this.(Module).getName().getText()
      or
      name = this.(Struct).getName().getText()
      or
      name = this.(Trait).getName().getText()
      or
      name = this.(Union).getName().getText()
      or
      this instanceof Use and
      name = "(use)"
    }

    Visibility getVisibility() {
      result = this.(Const).getVisibility()
      or
      result = this.(Enum).getVisibility()
      or
      result = this.(Function).getVisibility()
      or
      result = this.(Module).getVisibility()
      or
      result = this.(Struct).getVisibility()
      or
      result = this.(Trait).getVisibility()
      or
      result = this.(Union).getVisibility()
      or
      result = this.(Use).getVisibility()
    }

    bindingset[this]
    pragma[inline_late]
    predicate isPublic() { exists(this.getVisibility()) }

    string getName() { result = name }
  }

  pragma[nomagic]
  private Element getAnItemDescendant(NamedItem item) {
    getImmediateParent(result) = item
    or
    exists(Element mid |
      mid = getAnItemDescendant(item) and
      getImmediateParent(result) = mid and
      not mid instanceof NamedItem
    )
  }

  pragma[nomagic]
  private NamedItem getImmediateParentItem(NamedItem item) { item = getAnItemDescendant(result) }

  pragma[nomagic]
  private NamedItem getItemAncestorNonModule(NamedItem item) {
    (
      result = getImmediateParentItem(item)
      or
      exists(NamedItem mid |
        mid = getItemAncestorNonModule(item) and
        result = getImmediateParentItem(mid) and
        not mid instanceof Module
      )
    )
  }

  pragma[nomagic]
  private Module getImmediateParentModule(NamedItem item) {
    result = getItemAncestorNonModule(item)
  }

  private predicate sourceFileEdge(SourceFile f, NamedItem item) { item = f.getAnItem() }

  pragma[nomagic]
  private predicate sourceFileEdge(SourceFile f, string name, NamedItem item) {
    (
      sourceFileEdge(f, item)
      or
      exists(Use use |
        sourceFileEdge(f, _, use) and
        itemEdge(use, item)
      )
    ) and
    name = item.getName()
  }

  pragma[nomagic]
  private predicate itemEdge(NamedItem item1, NamedItem item2) {
    item1 = getImmediateParentItem(item2)
    or
    fileImportEdge(item1, item2)
    or
    useImportEdge(item1, item2)
  }

  pragma[nomagic]
  private predicate itemEdge(NamedItem item1, string name, NamedItem item2) {
    (
      itemEdge(item1, item2)
      or
      exists(Use use |
        itemEdge(item1, _, use) and
        itemEdge(use, item2)
      )
    ) and
    name = item2.getName()
    or
    name = "super" and
    if item1 instanceof Module
    then item2 = getImmediateParentModule(item1)
    else item2 = getImmediateParentModule(getImmediateParentModule(item1))
  }

  /** Holds if `f` is available as `mod name;` inside `folder`. */
  private predicate fileModule(SourceFile f, string name, Folder folder) {
    exists(File file | file = f.getFile() |
      file.getBaseName() = name + ".rs" and
      folder = file.getParentContainer()
      or
      exists(Folder encl |
        file.getBaseName() = "mod.rs" and
        encl = file.getParentContainer() and
        name = encl.getBaseName() and
        folder = encl.getParentContainer()
      )
    )
  }

  /** Holds if `m` is a `mod name;` item importing file `f`. */
  private predicate fileImport(Module m, SourceFile f) {
    exists(string name |
      not m.hasItemList() and
      // TODO: handle
      // ```
      // #[path = "foo.rs"]
      // mod bar;
      // ```
      not m.getAnAttr().getMeta().getPath().getPart().getNameRef().getText() = "path" and
      name = m.getName().getText() and
      fileModule(f, name, m.getFile().getParentContainer())
    )
  }

  /**
   * Holds if `mod` is a `mod name;` item resulting in the item path
   * `head::modPath` being valid. The new path is valid because
   * `filePath` is valid within the file that gets imported.
   */
  private predicate fileImportEdge(Module mod, NamedItem item) {
    item.isPublic() and
    exists(SourceFile f |
      fileImport(mod, f) and
      sourceFileEdge(f, item)
    )
  }

  pragma[nomagic]
  private predicate rootPath(Path path, string name, SourceFile f, NamedItem scope) {
    exists(NamedItem encl |
      f.getAnItem() = getImmediateParentItem*(scope) and
      getAnItemDescendant(encl) = path and
      scope = [encl, getImmediateParentModule(encl).(NamedItem)] and
      not exists(path.getQualifier()) and
      not path = any(UseTreeList list).getAUseTree().getPath() and // todo
      name = path.getPart().getNameRef().getText()
    )
  }

  pragma[nomagic]
  Item resolveItem(Path path) {
    exists(SourceFile f, NamedItem encl, string name | rootPath(path, name, f, encl) |
      itemEdge(encl, name, result)
      or
      sourceFileEdge(f, name, result)
    )
    or
    exists(Item q |
      q = resolveItem(path.getQualifier()) and
      itemEdge(q, path.getPart().getNameRef().getText(), result)
    )
  }

  /** Holds if `m` is a `mod name;` item importing file `f`. */
  pragma[nomagic]
  private predicate useImportEdge(Use use, NamedItem item) {
    // imported item must be `pub`
    item.isPublic() and
    exists(UseTree tree, Path path, NamedItem used |
      tree = use.getUseTree() and
      path = tree.getPath() and
      used = resolveItem(path) and
      // todo: handle `getUseTreeList`
      not exists(tree.getUseTreeList()) and
      if used instanceof Module
      then
        // glob import
        itemEdge(used, item)
      else item = used
    )
  }
  // /** A crate. */
  // class Crate extends File {
  //   Crate() { this.getBaseName() = "Cargo.toml" }
  //   pragma[nomagic]
  //   Item getItem() {
  //     exists(SourceFile file |
  //       file.getFile().getParentContainer() = this.getParentContainer() and
  //       file.getAnItem() = result
  //     )
  //   }
  // }
}
