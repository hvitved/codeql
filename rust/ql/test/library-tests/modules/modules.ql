import rust
import codeql.rust.elements.internal.ItemImpl::Impl as ItemImpl

query predicate mod(Module m) { any() }

query predicate resolveItem(Path p, Item i) { i = ItemImpl::resolveItem(p) }
