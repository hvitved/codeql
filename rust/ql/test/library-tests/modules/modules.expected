mod
| lib.rs:1:1:1:7 | mod my |
| main.rs:1:1:1:7 | mod my |
| main.rs:7:1:7:8 | mod my2 |
| main.rs:13:1:29:1 | mod m1 |
| main.rs:18:5:28:5 | mod m2 |
| my2/mod.rs:1:1:1:16 | mod nested2 |
| my2/nested2.rs:1:1:11:1 | mod nested3 |
| my2/nested2.rs:2:5:10:5 | mod nested4 |
| my.rs:1:1:1:15 | mod nested |
| my/nested.rs:1:1:17:1 | mod nested1 |
| my/nested.rs:2:5:11:5 | mod nested2 |
resolveItem
| main.rs:3:5:3:6 | my | main.rs:1:1:1:7 | mod my |
| main.rs:5:5:5:6 | my | main.rs:1:1:1:7 | mod my |
| main.rs:5:5:5:14 | ...::nested | my.rs:1:1:1:15 | mod nested |
| main.rs:9:5:9:7 | my2 | main.rs:7:1:7:8 | mod my2 |
| main.rs:11:5:11:7 | my2 | main.rs:7:1:7:8 | mod my2 |
| main.rs:11:5:11:16 | ...::nested2 | my2/mod.rs:1:1:1:16 | mod nested2 |
| main.rs:11:5:11:25 | ...::nested3 | my2/nested2.rs:1:1:11:1 | mod nested3 |
| main.rs:11:5:11:34 | ...::nested4 | my2/nested2.rs:2:5:10:5 | mod nested4 |
| main.rs:25:13:25:13 | f | main.rs:19:9:21:9 | fn f |
| main.rs:25:13:25:13 | f | my.rs:5:1:7:1 | fn f |
| main.rs:26:13:26:17 | super | main.rs:13:1:29:1 | mod m1 |
| main.rs:26:13:26:20 | ...::f | main.rs:14:5:16:5 | fn f |
| main.rs:41:17:41:19 | Foo | main.rs:31:1:31:13 | struct Foo |
| main.rs:41:17:41:19 | Foo | main.rs:39:9:39:21 | struct Foo |
| main.rs:44:13:44:15 | Foo | main.rs:31:1:31:13 | struct Foo |
| main.rs:44:13:44:15 | Foo | main.rs:36:5:36:17 | struct Foo |
| main.rs:46:5:46:5 | f | main.rs:38:5:42:5 | fn f |
| main.rs:46:5:46:5 | f | my.rs:5:1:7:1 | fn f |
| main.rs:50:13:50:15 | Foo | main.rs:31:1:31:13 | struct Foo |
| main.rs:54:5:54:6 | my | main.rs:1:1:1:7 | mod my |
| main.rs:54:5:54:14 | ...::nested | my.rs:1:1:1:15 | mod nested |
| main.rs:55:5:55:6 | my | main.rs:1:1:1:7 | mod my |
| main.rs:55:5:55:9 | ...::f | my.rs:5:1:7:1 | fn f |
| main.rs:56:5:56:11 | nested2 | my2/mod.rs:1:1:1:16 | mod nested2 |
| main.rs:56:5:56:20 | ...::nested3 | my2/nested2.rs:1:1:11:1 | mod nested3 |
| main.rs:56:5:56:29 | ...::nested4 | my2/nested2.rs:2:5:10:5 | mod nested4 |
| main.rs:56:5:56:32 | ...::f | my2/nested2.rs:3:9:5:9 | fn f |
| main.rs:57:5:57:5 | f | my.rs:5:1:7:1 | fn f |
| main.rs:59:5:59:5 | h | main.rs:33:1:47:1 | fn h |
| main.rs:59:5:59:5 | h | my.rs:9:1:12:1 | fn h |
| main.rs:60:5:60:6 | m1 | main.rs:13:1:29:1 | mod m1 |
| main.rs:60:5:60:10 | ...::m2 | main.rs:18:5:28:5 | mod m2 |
| main.rs:60:5:60:13 | ...::g | main.rs:23:9:27:9 | fn g |
| main.rs:61:5:61:5 | h | main.rs:33:1:47:1 | fn h |
| main.rs:61:5:61:5 | h | my.rs:9:1:12:1 | fn h |
| my2/mod.rs:5:5:5:11 | nested2 | my2/mod.rs:1:1:1:16 | mod nested2 |
| my2/mod.rs:5:5:5:20 | ...::nested3 | my2/nested2.rs:1:1:11:1 | mod nested3 |
| my2/mod.rs:5:5:5:29 | ...::nested4 | my2/nested2.rs:2:5:10:5 | mod nested4 |
| my2/mod.rs:5:5:5:32 | ...::f | my2/nested2.rs:3:9:5:9 | fn f |
| my.rs:3:5:3:10 | nested | my.rs:1:1:1:15 | mod nested |
| my/nested.rs:9:13:9:13 | f | my/nested.rs:3:9:5:9 | fn f |
| my/nested.rs:15:9:15:15 | nested2 | my/nested.rs:2:5:11:5 | mod nested2 |
| my/nested.rs:15:9:15:18 | ...::f | my/nested.rs:3:9:5:9 | fn f |
| my/nested.rs:21:5:21:11 | nested1 | my/nested.rs:1:1:17:1 | mod nested1 |
| my/nested.rs:21:5:21:20 | ...::nested2 | my/nested.rs:2:5:11:5 | mod nested2 |
| my/nested.rs:21:5:21:23 | ...::f | my/nested.rs:3:9:5:9 | fn f |
